### ISSUE TITLE
How to upload

### ISSUE THREAD

**Issue by MarvinNazari** [ishkawa/APIKit#103]
Is there anyway in APIKit that we can POST with a file (multpart/form-data)?
Sorry if i ask here

**Comment by MarvinNazari** [comment]
Would something like this work?

```swift
    public var requestBodyBuilder: RequestBodyBuilder {
        let boundary = NSString(format: "---------------------------V2ymHFg03esomerandomstuffhbqgZCaKO6jy")
        return .Custom(
            contentTypeHeader: "multipart/form-data; boundary=\(boundary)",
            buildBodyFromObject: { (image) -> NSData in
                
            let data = NSMutableData()
            if let imageData = UIImagePNGRepresentation(self.avatar) {
                
                data.appendData(NSString(format: "\r\n--%@\r\n", boundary).dataUsingEncoding(NSUTF8StringEncoding)!)
                data.appendData(NSString(format:"Content-Disposition: form-data; name=\"profile_photo\"; filename=\"img.jpg\"\\r\n").dataUsingEncoding(NSUTF8StringEncoding)!)
                data.appendData(NSString(format: "Content-Type: application/octet-stream\r\n\r\n").dataUsingEncoding(NSUTF8StringEncoding)!)
                data.appendData(imageData)
                data.appendData(NSString(format: "\r\n--%@\r\n", boundary).dataUsingEncoding(NSUTF8StringEncoding)!)
            }
            return data
        })
    }
```

**Comment by ishkawa** [comment]
Yes, that is a right way. The only mistake is that you've written extra `\` after `filename=\"img.jpg\"`. This causes invalid `\r` at the end of `Content-Disposition`.

```
POST /image HTTP/1.1
Host: localhost:1234
Accept: */*
Accept-Encoding: gzip, deflate
Accept-Language: en-us
Connection: keep-alive
Content-Type: multipart/form-data; boundary=---------------------------V2ymHFg03esomerandomstuffhbqgZCaKO6jy
User-Agent: xctest (unknown version) CFNetwork/758.1.6 Darwin/15.0.0


-----------------------------V2ymHFg03esomerandomstuffhbqgZCaKO6jy
Content-Disposition: form-data; name="profile_photo"; filename="img.jpg"\r
Content-Type: application/octet-stream

�PNG

$sRGB���	pHYs%%IR$�iDOT(CyeQ�IDATb�������!ok
                                                   IDATc�����G���d��IEND�B`�
-----------------------------V2ymHFg03esomerandomstuffhbqgZCaKO6jy
```

I used following server to confirm the raw request.

```go
package main

import (
	"fmt"
	"net/http"
	"net/http/httputil"
)

func ImageHandler(w http.ResponseWriter, r *http.Request) {
	bytes, err := httputil.DumpRequest(r, true)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	fmt.Println(string(bytes))

	w.WriteHeader(http.StatusOK)
}

func main() {
	http.HandleFunc("/image", ImageHandler)
	http.ListenAndServe(":1234", nil)
}
```

**Comment by MarvinNazari** [comment]
Yeah! that worked! thanks, but maybe we could make it somehow easier in APIKit to do Multpart stuff

**Comment by raulriera** [comment]
Hey guys,

I follow up on this thread, I am trying the following code

```swift
public var requestBodyBuilder: RequestBodyBuilder {
        let boundaryConstant = "---------------------------V2ymHFg03esASdsaffre3hbqgZCaKO6jy"
        let body = RequestBodyBuilder.Custom(
            contentTypeHeader: "multipart/form-data; boundary=\(boundaryConstant)",
            buildBodyFromObject: { _ in
                
                let uploadData = NSMutableData()
                
                if let imageData = UIImagePNGRepresentation(self.image) {
                    // add image
                    uploadData.appendData("\r\n--\(boundaryConstant)\r\n".dataUsingEncoding(NSUTF8StringEncoding)!)
                    uploadData.appendData("Content-Disposition: form-data; name=\"image\"; filename=\"file.png\"\r\n".dataUsingEncoding(NSUTF8StringEncoding)!)
                    uploadData.appendData("Content-Type: image/png\r\n\r\n".dataUsingEncoding(NSUTF8StringEncoding)!)
                    uploadData.appendData(imageData)
                    
                    // add parameters
                    for (key, value) in self.parameters {
                        uploadData.appendData("\r\n--\(boundaryConstant)\r\n".dataUsingEncoding(NSUTF8StringEncoding)!)
                        uploadData.appendData("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n\(value)".dataUsingEncoding(NSUTF8StringEncoding)!)
                    }
                    uploadData.appendData("\r\n--\(boundaryConstant)--\r\n".dataUsingEncoding(NSUTF8StringEncoding)!)
                }
                
                return uploadData
        })
        
        return body
    }
```

But I am getting this error from APIKit

```
.Failure(ResponseBodyDeserializationError(Error Domain=NSCocoaErrorDomain Code=3840 "JSON text did not start with array or object and option to allow fragments not set." UserInfo={NSDebugDescription=JSON text did not start with array or object and option to allow fragments not set.}))
```

Does anything pop into mind what could be causing this?

**Comment by raulriera** [comment]
Probably because Dribbble responds with an empty response? Digging a bit deeper here

**Comment by ishkawa** [comment]
Now, `RequestBodyBuilder. MultipartFormData` is available! Check out #128 for more details.
